hamilton/lifecycle/base.py

```diff
@@ -751,19 +751,21 @@ def _get_lifecycle_validators(
     def _get_lifecycle_hooks(
         self,
     ) -> Tuple[Dict[str, List[LifecycleAdapter]], Dict[str, List[LifecycleAdapter]]]:
-        sync_hooks = collections.defaultdict(set)
-        async_hooks = collections.defaultdict(set)
+        sync_hooks = collections.defaultdict(list)
+        async_hooks = collections.defaultdict(list)
         for adapter in self.adapters:
             for cls in inspect.getmro(adapter.__class__):
                 sync_hook = getattr(cls, SYNC_HOOK, None)
                 if sync_hook is not None:
-                    sync_hooks[sync_hook].add(adapter)
+                    if adapter not in sync_hooks[sync_hook]:
+                        sync_hooks[sync_hook].append(adapter)
```

Comment:
OK, this is not ironed into the contract, and we'll likely want to switch around order, but I think for the future we need to think about this a bit -- my instinct says to do `pre_` in order and `post_` in reverse order, but that's complicated and ambiguous to implement.

Reply:
yes we should do it that way ideally